pkg/phlaredb/schemas/v1/profiles.go

```diff
@@ -124,3 +127,242 @@ func (*ProfilePersister) Reconstruct(row parquet.Row) (id uint64, s *Profile, er
+	DropFrames int64
+	// frames with Function.function_name fully matching the following
+	// regexp will be kept, even if it matches drop_frames.
+	KeepFrames int64
+	// Time of collection (UTC) represented as nanoseconds past the epoch.
+	TimeNanos int64
+	// Duration of the profile, if a duration makes sense.
+	DurationNanos int64
+	// The number of events between sampled occurrences.
+	Period int64
+	// Freeform text associated to the profile.
+	Comments []int64
+	// Index into the string table of the type of the preferred sample
+	// value. If unset, clients should default to the last sample value.
+	DefaultSampleType int64
+
+	Samples Samples
+}
+
+type Samples struct {
+	StacktraceIDs []uint32
+	Values        []uint64
+}
+
+// Compact zero samples and optionally duplicates.
+func (s Samples) Compact(dedupe bool) Samples {
+	if len(s.StacktraceIDs) == 0 {
+		return s
+	}
+	if dedupe {
+		s = trimDuplicateSamples(s)
+	}
+	s = trimZeroSamples(s)
+	return s
+}
+
+func (s Samples) Clone() Samples {
+	return cloneSamples(s)
+}
+
+func trimDuplicateSamples(samples Samples) Samples {
+	sort.Sort(samples)
+	n := 0
+	for j := 1; j < len(samples.StacktraceIDs); j++ {
+		if samples.StacktraceIDs[n] == samples.StacktraceIDs[j] {
+			samples.Values[n] += samples.Values[j]
+		} else {
+			n++
+			samples.StacktraceIDs[n] = samples.StacktraceIDs[j]
+			samples.Values[n] = samples.Values[j]
+		}
+	}
+	return Samples{
+		StacktraceIDs: samples.StacktraceIDs[:n+1],
+		Values:        samples.Values[:n+1],
+	}
+}
+
+func trimZeroSamples(samples Samples) Samples {
+	n := 0
+	for j, v := range samples.Values {
+		if v != 0 {
+			samples.Values[n] = v
+			samples.StacktraceIDs[n] = samples.StacktraceIDs[j]
+			n++
+		}
+	}
+	return Samples{
+		StacktraceIDs: samples.StacktraceIDs[:n],
+		Values:        samples.Values[:n],
+	}
+}
+
+func cloneSamples(samples Samples) Samples {
+	return Samples{
+		StacktraceIDs: copySlice(samples.StacktraceIDs),
+		Values:        copySlice(samples.Values),
+	}
+}
+
+func (s Samples) Less(i, j int) bool {
+	return s.StacktraceIDs[i] < s.StacktraceIDs[j]
+}
+
+func (s Samples) Swap(i, j int) {
+	s.StacktraceIDs[i], s.StacktraceIDs[j] = s.StacktraceIDs[j], s.StacktraceIDs[i]
+	s.Values[i], s.Values[j] = s.Values[j], s.Values[i]
+}
+
+func (s Samples) Len() int {
+	return len(s.StacktraceIDs)
+}
+
+func (p InMemoryProfile) Timestamp() model.Time {
+	return model.TimeFromUnixNano(p.TimeNanos)
+}
+
+func (p InMemoryProfile) Total() int64 {
+	var total int64
+	for _, sample := range p.Samples.Values {
+		total += int64(sample)
+	}
+	return total
+}
+
+func copySlice[T any](in []T) []T {
+	out := make([]T, len(in))
+	copy(out, in)
+	return out
+}
+
+func NewInMemoryProfilesRowReader(slice []InMemoryProfile) *SliceRowReader[InMemoryProfile] {
+	return &SliceRowReader[InMemoryProfile]{
+		slice:     slice,
+		serialize: DeconstructMemoryProfile,
+	}
+}
+
+func DeconstructMemoryProfile(imp InMemoryProfile, row parquet.Row) parquet.Row {
```

Comment:
Maybe worth keep scope to the package:

```suggestion
func deconstructMemoryProfile(imp InMemoryProfile, row parquet.Row) parquet.Row {
```